src/basic/fileio.c

```diff
@@ -856,35 +857,41 @@ int read_full_file_full(
         return read_full_stream_full(f, filename, offset, size, flags, ret_contents, ret_size);
 }
 
-int executable_is_script(const char *path, char **interpreter) {
-        _cleanup_free_ char *line = NULL;
-        size_t len;
-        char *ans;
+int script_get_interpreter(const char *path, char **ret) {
+        _cleanup_fclose_ FILE *f = NULL;
         int r;
 
         assert(path);
 
-        r = read_one_line_file(path, &line);
-        if (r == -ENOBUFS) /* First line overly long? if so, then it's not a script */
-                return 0;
-        if (r < 0)
-                return r;
+        f = fopen(path, "re");
+        if (!f)
+                return -errno;
 
-        if (!startswith(line, "#!"))
-                return 0;
+        char buf[STRLEN("#!")];
+        if (fread(buf, 1, sizeof(buf), f) != sizeof(buf)) {
+                if (feof(f))
+                        return -EINVAL;
+
+                return -EIO;
+        }
+        if (!strneq(buf, "#!", sizeof(buf)))
+                return -ENOTTY;
```

Comment:
i'd return the same error here as if we hit EOF, i.e. EBADMSG too, as suggested above

Reply:
Hmm, I figure ENOTTY is more appropriate here, i.e. we didn't get a script, but something else.

Reply:
ENOTTY is recognized by ERRNO_IS_NOT_SUPPORTED() as a form of "not supported", but i think a file that doesn't have a shebang line isn't "not support". it's just not a script...

Reply:
ENOEXEC?